File: src/main/java/logic/MainLogic.java
package logic;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

import lib.ConfigurableOption;
import lib.DrawingUtility;
import lib.GameAnimation;
import lib.GameBackground;
import lib.GameManager;
import lib.HighScoreUtility;
import lib.IGameLogic;
import lib.IRenderableHolder;
import lib.IRenderableObject;
import lib.InputUtility;
import lib.RandomUtility;

public class MainLogic implements IRenderableHolder,IGameLogic{

	//All renderable objects
	private GameBackground background;
	private PlayerStatus player;
	private List<TargetObject> onScreenObject = new ArrayList<TargetObject>();
	private List<GameAnimation> onScreenAnimation = new ArrayList<GameAnimation>();
	
	/*
	 * Reserved z
	 * MIN_VALUE : background
	 * MAX_VALUE-1 : animation effect
	 * MAX_VALUE : player's status
	 */
	private int zCounter = Integer.MIN_VALUE+1;
	private int nextObjectCreationDelay;	
	private boolean readyToRender = false; //For dealing with synchronization issue
	
	//Called before enter the game loop
	public synchronized void onStart(){
		background = new GameBackground();
		player = new PlayerStatus();
		player.setCurrentGun(new Gun(1));
		nextObjectCreationDelay = RandomUtility.random(ConfigurableOption.objectCreationMinDelay, 
				ConfigurableOption.objectCreationMaxDelay);
		readyToRender = true;
	}
	
	//Called after exit the game loop
	public synchronized void onExit(){
		readyToRender = false;
		onScreenObject.clear();
	}
	
	public void logicUpdate(){
		//Paused
		if(InputUtility.isTypeEnter()){
			/* fill code1 */
		}
		
		if(player.isPause()){
			return;
		}
		
		//Update moving background
		background.updateBackground();
		
		//Time up
		if(player.getRemainingTime() == 0){
			HighScoreUtility.recordHighScore(player.getScore());
			GameManager.goToTitle();
			return;
		}
		player.decreaseRemainingTime(1);
		
		//Create random target
		createTarget();
		
		//Shoot and grab
		TargetObject target = null;
		TargetObject grabbedObject = null;
		if(!player.isDisplayingArea(InputUtility.getMouseX(), InputUtility.getMouseY())){
			
			boolean shoot = false;
			if((InputUtility.isMouseLeftClicked() || InputUtility.isTypeSpace()) && 
					player.getCurrentGun() != null &&
					player.getCurrentGun().canShoot()){
				
				/* fill code2 */
								
				onScreenAnimation.add(DrawingUtility.createShootingAnimationAt(
						InputUtility.getMouseX(), InputUtility.getMouseY()));
				shoot = true;
			}
			
			target = getTopMostTargetAt(InputUtility.getMouseX(), InputUtility.getMouseY());
			if(target != null){
				
				/* fill code3 */				
			}
			
			if(shoot && player.getCurrentGun() instanceof SpecialGun && 
					((SpecialGun)player.getCurrentGun()).getBulletQuantity() <= 0){
				
				/* fill code4 */				
			}
		}
		
		//Update target object
		for(TargetObject obj : onScreenObject){
			if(obj instanceof CollectibleObject && grabbedObject != obj){
				((CollectibleObject)obj).ungrab(); //Ungrab
			}
			obj.move();
		}
		
		//Update animation
		for(GameAnimation obj : onScreenAnimation){
			obj.updateAnimation();
		}
		
		//Remove unused image
		for(int i=onScreenObject.size()-1; i>=0; i--){
			if(onScreenObject.get(i).isDestroyed)
				onScreenObject.remove(i);
		}
		for(int i=onScreenAnimation.size()-1; i>=0; i--){
			if(!onScreenAnimation.get(i).isVisible())
				onScreenAnimation.remove(i);
		}
	}
	
	private void createTarget(){
		if(nextObjectCreationDelay > 0){
			nextObjectCreationDelay--;
		}else{
			//Random next creation delay
			nextObjectCreationDelay = RandomUtility.random(ConfigurableOption.objectCreationMinDelay, 
					ConfigurableOption.objectCreationMaxDelay);
			
			//Random moving duration
			int movingDuration = RandomUtility.random(ConfigurableOption.objectMinDuration, ConfigurableOption.objectMaxDuration);
			
			//Random target type
			/* fill code */
			
			//Increase z counter (so the next object will be created on top of the previous one)
			zCounter++;
			if(zCounter == Integer.MAX_VALUE-1){
				zCounter = Integer.MIN_VALUE+1;
			}
		}
	}
	
	private TargetObject getTopMostTargetAt(int x,int y){
		TargetObject obj = null;
		for(TargetObject target : onScreenObject){
			if(target.contains(x, y)){
				if(obj != null){
					if(target.getZ() > obj.getZ()){
						obj.setPointerOver(false);
						obj = target;
						obj.setPointerOver(true);
					}
				}else{
					obj = target;
					obj.setPointerOver(true);
				}
			}else{
				target.setPointerOver(false);
			}
		}
		return obj;
	}

	@Override
	public synchronized List<IRenderableObject> getSortedRenderableObject() {
		List<IRenderableObject> sortedRenderable = new ArrayList<IRenderableObject>();
		if(!readyToRender) return sortedRenderable;
		for(TargetObject object : onScreenObject){
			sortedRenderable.add(object);
		}
		for(GameAnimation object : onScreenAnimation){
			sortedRenderable.add(object);
		}
		sortedRenderable.add(player);
		sortedRenderable.add(background);
		
		Collections.sort(sortedRenderable, new Comparator<IRenderableObject>() {
			@Override
			public int compare(IRenderableObject o1, IRenderableObject o2) {
				if(o1.getZ() > o2.getZ())
					return 1;
				else if(o1.getZ() < o2.getZ())
					return -1;
				else
					return 0;
			}
		});
		return sortedRenderable;
	}
}

File: src/logic/MainLogic.kt
package logic

import java.util.ArrayList
import java.util.Collections

import lib.ConfigurableOption
import lib.DrawingUtility
import lib.GameAnimation
import lib.GameBackground
import lib.GameManager
import lib.HighScoreUtility
import lib.IGameLogic
import lib.IRenderableHolder
import lib.IRenderableObject
import lib.InputUtility
import lib.RandomUtility

class MainLogic : IRenderableHolder, IGameLogic {
    /* all renderable objects */
    lateinit private var background: GameBackground
    lateinit private var player: PlayerStatus
    private val onScreenObject = mutableListOf<TargetObject>()
    private val onScreenAnimation = mutableListOf<GameAnimation>()

    /*
	 * Reserved z MIN_VALUE : background MAX_VALUE-1 : animation effect
	 * MAX_VALUE : player's status
	 */
    private var zCounter = Int.MIN_VALUE + 1
    private var nextObjectCreationDelay = 0
    private var readyToRender = false /* for dealing with synchronization issue */

    /* called before entering the game loop */
    @Synchronized override fun onStart() {
        background = GameBackground()
        player = PlayerStatus()
        player.currentGun = Gun(1)
        nextObjectCreationDelay = RandomUtility.random(ConfigurableOption.objectCreationMinDelay, ConfigurableOption.objectCreationMaxDelay)
        readyToRender = true
    }

    /* called after exiting the game loop */
    @Synchronized override fun onExit() {
        readyToRender = false
        onScreenObject.clear()
    }

    override fun logicUpdate() {
        if (InputUtility.isTypeEnter())
            player.pause = !player.pause

        if (player.pause)
            return

        /* update moving background */
        background.updateBackground()

        if (player.remainingTime == 0) {
            HighScoreUtility.recordHighScore(player.score)
            GameManager.goToTitle()
            return
        }
        player.decreaseRemainingTime(1)

        /* create random target */
        createTarget()

        /* shoot and grab */
        val target: TargetObject?
        var grabbedObject: TargetObject? = null
        if (!player.isDisplayingArea(InputUtility.getMouseX(), InputUtility.getMouseY())) {
            var shoot = false
            if ((InputUtility.isMouseLeftClicked() || InputUtility.isTypeSpace()) && player.currentGun.canShoot()) {
                player.currentGun.shoot()
                onScreenAnimation.add(DrawingUtility.createShootingAnimationAt(InputUtility.getMouseX(), InputUtility.getMouseY()))
                shoot = true
            }
            target = getTopMostTargetAt(InputUtility.getMouseX(), InputUtility.getMouseY())
            if (target is CollectibleObject) {
                target.grab(player)
                grabbedObject = target
            } else if (shoot && target is ShootableObject) {
                target.hit(player)
            }

            if (shoot && player.currentGun is SpecialGun && (player.currentGun as SpecialGun).bulletQuantity <= 0) {
                player.currentGun = Gun(1)
            }
        }
        /* update objects */
        onScreenObject.filter { it !== grabbedObject }.filterIsInstance<CollectibleObject>().forEach { it.ungrab() }
        onScreenObject.forEach { it.move() }

        /* update animations */
        onScreenAnimation.forEach { it.updateAnimation() }

        /* remove unused images */
        onScreenObject.filter { it.isDestroyed }.forEach { onScreenObject.remove(it) }
        onScreenAnimation.filter { !it.isVisible }.forEach { onScreenAnimation.remove(it) }
    }

    private fun createTarget() {
        if (nextObjectCreationDelay > 0) {
            nextObjectCreationDelay--
        } else {
            /* random next creation delay */
            nextObjectCreationDelay = RandomUtility.random(ConfigurableOption.objectCreationMinDelay, ConfigurableOption.objectCreationMaxDelay)

            /* random moving duration */
            val movingDuration = RandomUtility.random(ConfigurableOption.objectMinDuration, ConfigurableOption.objectMaxDuration)

            /* random target type */
            val rand = RandomUtility.random(1, 100)
            when(rand) {
                in 1..15 -> onScreenObject.add(if (player.currentGun is SpecialGun) ItemBullet(movingDuration, zCounter) else ItemSpecialGun(movingDuration, zCounter))
                in 16..35 -> onScreenObject.add(SplitterTarget(40, movingDuration, zCounter, onScreenObject))
                in 36..70 -> onScreenObject.add(SmallTarget(15, movingDuration, zCounter))
                else -> onScreenObject.add(SimpleTarget(15, movingDuration, zCounter))
            }

            /* Increase z counter (so the next object will be created on top of the previous one) */
            zCounter++
            if (zCounter == Integer.MAX_VALUE - 1) {
                zCounter = Integer.MIN_VALUE + 1
            }
        }
    }

    private fun getTopMostTargetAt(x: Int, y: Int): TargetObject? {
        onScreenObject.filterNot { it.contains(x,y) }.forEach { it.isPointerOver = false }
        val obj = onScreenObject.filter { it.contains(x,y) }.maxBy { it.z }
        obj?.isPointerOver = true
        return obj
    }

    @Synchronized override fun getSortedRenderableObject(): List<IRenderableObject> {
        val sortedRenderable = ArrayList<IRenderableObject>()
        if (!readyToRender)
            return sortedRenderable
        onScreenObject.forEach { sortedRenderable.add(it) }
        onScreenAnimation.forEach { sortedRenderable.add(it) }
        sortedRenderable.add(player)
        sortedRenderable.add(background)

        Collections.sort(sortedRenderable) { o1, o2 ->
            if (o1.z > o2.z)
                1
            else if (o1.z < o2.z)
                -1
            else
                0
        }
        return sortedRenderable
    }
}
